src/neural/network_blas.cc

```diff
@@ -17,270 +17,314 @@
+                     weights_.input.bn_means.data(),
+                     weights_.input.bn_stddivs.data());
 
     // Residual tower
-    auto conv_in = std::vector<float>(output_channels * width * height);
-    auto res = std::vector<float>(output_channels * width * height);
 
     for (auto& residual : weights_.residual) {
       auto& conv1 = residual.conv1;
-      auto output_channels = conv1.biases.size();
+      auto& conv2 = residual.conv2;
+
       std::swap(conv_out, conv_in);
-      std::copy(begin(conv_in), end(conv_in), begin(res));
 
-      Transforms::WinogradConvolve3(output_channels, conv_in, conv1.weights, V,
-                                    M, conv_out);
-      Transforms::Batchnorm<64>(output_channels, conv_out,
-                                conv1.bn_means.data(), conv1.bn_stddivs.data());
+      convolve3.Forward(batch_size, output_channels, output_channels, conv_in,
+                        &conv1.weights[0], conv_out);
 
-      auto& conv2 = residual.conv2;
-      output_channels = conv2.biases.size();
+      Batchnorm::Apply(batch_size, output_channels, &conv_out[0],
+                       conv1.bn_means.data(), conv1.bn_stddivs.data());
+
+      std::swap(conv_in, res);
       std::swap(conv_out, conv_in);
-      Transforms::WinogradConvolve3(output_channels, conv_in, conv2.weights, V,
-                                    M, conv_out);
-      Transforms::Batchnorm<64>(output_channels, conv_out,
-                                conv2.bn_means.data(), conv2.bn_stddivs.data(),
-                                res.data());
-    }
 
-    Transforms::Convolve<1>(NUM_POLICY_INPUT_PLANES, conv_out,
-                            weights_.policy.weights, weights_.policy.biases,
-                            policy_data);
+      convolve3.Forward(batch_size, output_channels, output_channels, conv_in,
+                        &conv2.weights[0], conv_out);
 
-    Transforms::Convolve<1>(NUM_VALUE_INPUT_PLANES, conv_out,
-                            weights_.value.weights, weights_.value.biases,
-                            value_data);
+      Batchnorm::Apply(batch_size, output_channels, conv_out,
+                       conv2.bn_means.data(), conv2.bn_stddivs.data(), res);
+    }
 
-    Transforms::Batchnorm<width * height>(NUM_POLICY_INPUT_PLANES, policy_data,
-                                          weights_.policy.bn_means.data(),
-                                          weights_.policy.bn_stddivs.data());
+    Convolution<1>::Forward(
+        batch_size, output_channels, num_policy_input_planes, conv_out,
+        weights_.policy.weights.data(), weights_.policy.biases.data(),
+        policy_buffer.data());
 
-    Transforms::Batchnorm<width * height>(NUM_VALUE_INPUT_PLANES, value_data,
-                                          weights_.value.bn_means.data(),
-                                          weights_.value.bn_stddivs.data());
+    Convolution<1>::Forward(batch_size, output_channels, num_value_input_planes,
```

Comment:
The template argument is *always* `1`, correct?

Reply:
Please leave the general discussions (pointer use and convolve<> code suppression) out of the line-by-line review, (except when it makes sense of course).